vnext/JSI/Shared/ChakraObjectRef.cpp

```diff
@@ -0,0 +1,325 @@
+  return ChakraObjectRef(result);
+
+#else
+  std::wstring utf16 = Common::Unicode::Utf8ToUtf16(utf8, length);
+  return ToJsString(utf16.c_str(), utf16.length());
+#endif
+}
+
+ChakraObjectRef ToJsString(const wchar_t *const utf16, size_t length) {
+  assert(utf16);
+  JsValueRef result = nullptr;
+  ThrowUponChakraError(
+      JsPointerToString(utf16, length, &result), "JsPointerToString");
+  return ChakraObjectRef(result);
+}
+
+ChakraObjectRef ToJsString(const ChakraObjectRef &ref) {
+  JsValueRef str = nullptr;
+  ThrowUponChakraError(
+      JsConvertValueToString(ref, &str), "JsConvertValueToString");
+  return ChakraObjectRef(str);
+}
+
+ChakraObjectRef ToJsNumber(int num) {
+  JsValueRef result = nullptr;
+  ThrowUponChakraError(JsIntToNumber(num, &result), "JsIntToNumber");
+  return ChakraObjectRef(result);
+}
+
+ChakraObjectRef ToJsArrayBuffer(
+    const std::shared_ptr<const facebook::jsi::Buffer> &buffer) {
+  assert(buffer);
+
+  size_t size = buffer->size();
+  assert(size < UINT_MAX);
+
+  JsValueRef arrayBuffer = nullptr;
+  auto bufferWrapper =
+      std::make_unique<std::shared_ptr<const facebook::jsi::Buffer>>(buffer);
+
+  // We allocate a copy of buffer on the heap, a shared_ptr which is deleted
+  // when the JavaScript garbage collecotr releases the created external array
+  // buffer. This ensures that buffer stays alive while the JavaScript engine is
+  // using it.
+  ThrowUponChakraError(
+      JsCreateExternalArrayBuffer(
+          Common::Utilities::CheckedReinterpretCast<char *>(
+              const_cast<uint8_t *>(buffer->data())),
+          static_cast<unsigned int>(size),
+          [](void *bufferToDestroy) {
+            // We wrap bufferToDestroy in a unique_ptr to avoid calling delete
+            // explicitly.
+            std::unique_ptr<std::shared_ptr<const facebook::jsi::Buffer>>
+                wrapper{
+                    static_cast<std::shared_ptr<const facebook::jsi::Buffer> *>(
+                        bufferToDestroy)};
+          },
+          bufferWrapper.get(),
+          &arrayBuffer),
+      "JsCreateExternalArrayBuffer");
+
+  bufferWrapper.release();
+  return ChakraObjectRef(arrayBuffer);
+}
+
+bool CompareJsValues(
+    const ChakraObjectRef &jsValue1,
+    const ChakraObjectRef &jsValue2) {
+  bool result = false;
+  // Note that JsStrictEquals should only be used for JsValueRefs and not for
+  // other types of JsRefs (e.g. JsPropertyIdRef, etc.).
+  ThrowUponChakraError(
+      JsStrictEquals(jsValue1, jsValue2, &result), "JsStrictEquals");
+  return result;
+}
+
+bool CompareJsPropertyIds(
+    const ChakraObjectRef &jsPropId1,
+    const ChakraObjectRef &jsPropId2) {
+  JsPropertyIdType type1 = GetPropertyIdType(jsPropId1);
+  JsPropertyIdType type2 = GetPropertyIdType(jsPropId2);
+
+  if (type1 != type2) {
+    return false;
+  }
+
+  if (type1 == JsPropertyIdTypeString) {
+    assert(type2 == JsPropertyIdTypeString);
```

Comment:
>assert(type2 == JsPropertyIdTypeString); [](start = 4, length = 40)

this check is not needed right? It's already done in 306

Same for line 316

Reply:
You are right. I'll remove them.

---
In reply to: [336176400](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336176400) [](ancestors = 336176400)